### PR TITLE
Enable NWC to mutate workflows created by service accounts

### PR DESCRIPTION
**Pull request by iameskild**
## Reference Issues or PRs

This PR is a related to the work being done to integrate Jupyter-Scheduler ([#1832](https://github.com/nebari-dev/nebari/pull/1832)).

## What does this implement/fix?

_Put a `x` in the boxes that apply_

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [ ] Did you test the pull request locally?
- [ ] Did you add new tests?

#### Text styling

- [ ] The content is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [ ] If there are headers, they use the proper header tags (with only one level-one header: `H1` or `#` in markdown).
- [ ] All links describe where they link to (for example, check the [Nebari website](https://nebari.dev/)).
- [ ] This content adheres to the Nebari style guides.

#### Non-text content

- [ ] All content is represented as text (for example, images need alt text, and videos need captions or descriptive transcripts).
- [ ] If there are emojis, there are not more than three in a row.
- [ ] Don't use [flashing GIFs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If the content were to be read as plain text, it still makes sense, and no information is missing.

### CODE REVIEW

File: nebari_workflow_controller/utils.py [hunk 1]
```diff
@@ -45,6 +48,28 @@ def sent_by_argo(workflow: dict):
     return None
 
 
+def valid_argo_roles():
+    # TODO: determine a more extensible way of generating a list of valid roles
+    return ["argo-admin", "argo-developer"]
```
Comment (Adam-D-Lewis) [inline nebari_workflow_controller/utils.py:53]: I think this should be a config option somehow.  Most other things in Nebari use traitlets, but I'd feel okay using an VALID_ARGO_ROLES ENV var for now.

File: nebari_workflow_controller/utils.py [hunk 2]
```diff
@@ -88,8 +113,21 @@ def get_keycloak_uid_username(
...
+            logger.warning(
+                f"Keycloak user with UID `{keycloak_uid}` not found. Checking if workflow was created by system-serviceaccount..."
+            )
+            preferred_username = workflow["metadata"]["labels"][
+                "workflows.argoproj.io/creator-preferred-username"
+            ]
+            if validate_service_account(keycloak_uid):
+                for user in kcadm.get_users():
+                    if user["username"] == preferred_username:
+                        return user["id"], preferred_username
```
Comment (Adam-D-Lewis) [inline nebari_workflow_controller/utils.py:129]: This won't work in some cases. You can't always go back to the actual username b/c "Labels only contain [-_.0-9a-zA-Z]", so any other characters will be turned into "-") so not a 1 to 1 mapping.

File: nebari_workflow_controller/utils.py [hunk 3]
```diff
@@ -274,22 +312,24 @@ def mutate_template(
     spec_keep_portions,
     template,
 ):
-    for value, key in container_keep_portions:
-        if "container" not in template:
-            continue
+    target = "container" if "container" in template else "script" if "script" in template else None
```
Comment (Adam-D-Lewis) [inline nebari_workflow_controller/utils.py:315]: These are argo workflows template types. So I'd think we should have a variable called SUPPORTED_TEMPLATE_TYPES and then we check if each is in the template.
```
SUPPORTED_TEMPLATE_TYPES = {'container', 'script'}
```

**Review by Adam-D-Lewis — CHANGES_REQUESTED** [review]
see comments

File: nebari_workflow_controller/utils.py (hunk 1 above)
Comment (iameskild) [reply nebari_workflow_controller/utils.py]: Agreed, this is not a good solution. I added a `configmap` to the nebari-workflow-controller deployment that provides a list of valid Argo roles ([here](https://github.com/nebari-dev/nebari/pull/1832/commits/5e5aff2ae6d32360d60ffd8a922c8dc6d1244e0f)). How does that solution work for you?

**Comment by Adam-D-Lewis** [comment]
Have you run the tests?  Could we add a test for the new functionality?

**Review by costrouc — APPROVED** [review]
LGTM. But like @Adam-D-Lewis mentioned it would be helpful to have tests for this worklfow.

**Comment by iameskild** [comment]
@costrouc @Adam-D-Lewis I added new tests that cover the special case where the Workflow was created via a service-account.

**Review by costrouc — APPROVED** [review]
@iameskild this are some detailed tests and I like the fixtures that you created in `conftest.py`. Thanks! LGTM.

**Comment by costrouc** [comment]
Since @Adam-D-Lewis is on PTO and we are trying to get a release out for nebari. I am comfortable with merging. @iameskild feel free to merge when it works best.

**Pull request merged**